README.md
| index | [npm-run-all] | [run-s] | [run-p] | [Node API] |
|-------|---------------|---------|---------|------------|

# npm-run-all2

[![npm version](https://img.shields.io/npm/v/npm-run-all2.svg)](https://www.npmjs.com/package/npm-run-all2)
[![Downloads/month](https://img.shields.io/npm/dm/npm-run-all2.svg)](http://www.npmtrends.com/npm-run-all2)
[![Actions Status](https://github.com/bcomnes/npm-run-all2/workflows/test/badge.svg)](https://github.com/bcomnes/npm-run-all2/actions)
[![Coverage Status](https://codecov.io/gh/bcomnes/npm-run-all2/branch/master/graph/badge.svg)](https://codecov.io/gh/bcomnes/npm-run-all2)
[![Dependency Status](https://david-dm.org/bcomnes/npm-run-all2.svg)](https://david-dm.org/bcomnes/npm-run-all2)

A CLI tool to run multiple npm-scripts in parallel or sequential.

## npm-run-all2 why?

**A maintenance fork of npm-run-all.**  npm-run-all2 is a fork of npm-run-all with dependabot updates, release automation enabled and some troublesome babel stuff removed to further reduce maintenance burden.  Hopefully this labor can upstream some day when mysticatea returns, but until then, welcome aboard!

## ⤴️ Motivation

- **Simplify.** The official `npm run-script` command cannot run multiple scripts, so if we want to run multiple scripts, it's redundant a bit. Let's shorten it by glob-like patterns.<br>
  Before: `npm run clean && npm run build:css && npm run build:js && npm run build:html`<br>
  After: `npm-run-all clean build:*`
- **Cross platform.** We sometimes use `&` to run multiple command in parallel, but `cmd.exe` (`npm run-script` uses it by default) does not support the `&`. Half of Node.js users are using it on Windows, so the use of `&` might block contributions. `npm-run-all --parallel` works well on Windows as well.

## 💿 Installation

```bash
$ npm install npm-run-all --save-dev
# or
$ yarn add npm-run-all --dev
```

- It requires `Node@>=4`.

## 📖 Usage

### CLI Commands

This `npm-run-all` package provides 3 CLI commands.

- [npm-run-all]
- [run-s]
- [run-p]

The main command is [npm-run-all].
We can make complex plans with [npm-run-all] command.

Both [run-s] and [run-p] are shorthand commands.
[run-s] is for sequential, [run-p] is for parallel.
We can make simple plans with those commands.

#### Yarn Compatibility

If a script is invoked with Yarn, `npm-run-all` will correctly use Yarn to execute the plan's child scripts.

### Node API

This `npm-run-all` package provides Node API.

- [Node API]

## 📰 Changelog

- https://github.com/mysticatea/npm-run-all/releases

## 🍻 Contributing

Welcome♡

### Bug Reports or Feature Requests

Please use GitHub Issues.

### Correct Documents

Please use GitHub Pull Requests.

I'm not familiar with English, so I especially thank you for documents' corrections.

### Implementing

Please use GitHub Pull Requests.

There are some npm-scripts to help developments.

- **npm test** - Run tests and collect coverage.
- **npm run clean** - Delete temporary files.
- **npm run lint** - Run ESLint.
- **npm run watch** - Run tests (not collect coverage) on every file change.

[npm-run-all]: docs/npm-run-all.md
[run-s]: docs/run-s.md
[run-p]: docs/run-p.md
[Node API]: docs/node-api.md


CHANGELOG.md
# Changelog

All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/)
and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).

Generated by [`auto-changelog`](https://github.com/CookPete/auto-changelog).

## [v5.0.1](https://github.com/bcomnes/npm-run-all2/compare/v5.0.0...v5.0.1)

### Commits

- Fix repo field to a valid format [`00b88f8`](https://github.com/bcomnes/npm-run-all2/commit/00b88f8a399d45cb104a33357cf56015ab92a1c0)
- Remove duplicate repo field [`a2d11ff`](https://github.com/bcomnes/npm-run-all2/commit/a2d11ff3f234812ba660be32f3a9a0aa45a510f6)
- Update FUNDING.yml [`648a541`](https://github.com/bcomnes/npm-run-all2/commit/648a5418725b4330571e08e9e1300756c98edd76)

## [v5.0.0](https://github.com/bcomnes/npm-run-all2/compare/v4.1.5...v5.0.0) - 2020-10-04

### Merged

- report codecov [`#21`](https://github.com/bcomnes/npm-run-all2/pull/21)
- Use built in string.padEnd [`#22`](https://github.com/bcomnes/npm-run-all2/pull/22)
- Upgrade: Bump fs-extra from 7.0.1 to 9.0.1 [`#17`](https://github.com/bcomnes/npm-run-all2/pull/17)
- remove @types/node [`#20`](https://github.com/bcomnes/npm-run-all2/pull/20)
- Upgrade: Bump @types/node from 4.9.4 to 14.11.2 [`#18`](https://github.com/bcomnes/npm-run-all2/pull/18)
- remove dependency on chalk [`#19`](https://github.com/bcomnes/npm-run-all2/pull/19)
- Upgrade: Bump eslint-config-mysticatea from 12.0.0 to 13.0.2 [`#11`](https://github.com/bcomnes/npm-run-all2/pull/11)
- Upgrade: Bump rimraf from 2.7.1 to 3.0.2 [`#10`](https://github.com/bcomnes/npm-run-all2/pull/10)
- Upgrade: Bump ansi-styles from 3.2.1 to 4.2.1 [`#12`](https://github.com/bcomnes/npm-run-all2/pull/12)
- Upgrade: Bump cross-spawn from 6.0.5 to 7.0.3 [`#9`](https://github.com/bcomnes/npm-run-all2/pull/9)
- Enable windows testing [`#15`](https://github.com/bcomnes/npm-run-all2/pull/15)
- Upgrade: Bump babel-preset-power-assert from 2.0.0 to 3.0.0 [`#3`](https://github.com/bcomnes/npm-run-all2/pull/3)
- more renaming [`#13`](https://github.com/bcomnes/npm-run-all2/pull/13)
- Upgrade: Bump p-queue from 2.4.2 to 6.6.1 [`#4`](https://github.com/bcomnes/npm-run-all2/pull/4)
- Upgrade: Bump nyc from 11.9.0 to 15.1.0 [`#7`](https://github.com/bcomnes/npm-run-all2/pull/7)
- Upgrade: Bump pidtree from 0.3.1 to 0.5.0 [`#5`](https://github.com/bcomnes/npm-run-all2/pull/5)
- Upgrade: Bump read-pkg from 3.0.0 to 5.2.0 [`#6`](https://github.com/bcomnes/npm-run-all2/pull/6)
- Upgrade: Bump mocha from 5.2.0 to 8.1.3 [`#8`](https://github.com/bcomnes/npm-run-all2/pull/8)
- Upgrade: Bump actions/setup-node from v2.1.1 to v2.1.2 [`#1`](https://github.com/bcomnes/npm-run-all2/pull/1)
- Upgrade: Bump actions/checkout from v2.3.2 to v2.3.3 [`#2`](https://github.com/bcomnes/npm-run-all2/pull/2)
- 🐛 fix signal handling [`#171`](https://github.com/bcomnes/npm-run-all2/pull/171)

### Commits

- fix lint [`acce8f7`](https://github.com/bcomnes/npm-run-all2/commit/acce8f78b0a4ac656020c14ab946d57639e22764)
- ⚒ switch to actions and use dependabot [`13696f8`](https://github.com/bcomnes/npm-run-all2/commit/13696f8aef475fb66346a8b219c9f07d026d2513)
- Fix lint [`87b3e01`](https://github.com/bcomnes/npm-run-all2/commit/87b3e01c1405bb92c39a1aecdf957805bcc184c0)

## [v4.1.5](https://github.com/bcomnes/npm-run-all2/compare/v4.1.4...v4.1.5) - 2018-11-24

### Commits

- 🐛 use pidtree [`1b41ac5`](https://github.com/bcomnes/npm-run-all2/commit/1b41ac569987c96e224f940ff59f9699322c7824)
- ⬆️ update dependencies [`7ec542e`](https://github.com/bcomnes/npm-run-all2/commit/7ec542e95ceb922b9abe593270d9b6f8e0df4bc5)
- 🔖 4.1.5 [`df15118`](https://github.com/bcomnes/npm-run-all2/commit/df1511851a2b5e8a406e4a2622829b360f671afc)

## [v4.1.4](https://github.com/bcomnes/npm-run-all2/compare/v4.1.3...v4.1.4) - 2018-11-24

### Commits

- 🔥 remove ps-tree [`57d72eb`](https://github.com/bcomnes/npm-run-all2/commit/57d72eb98c2ce108f07d2a2cf1b44d57f08ec3ca)
- 🔖 4.1.4 [`a79fbac`](https://github.com/bcomnes/npm-run-all2/commit/a79fbacda181ec8a240d8bf8d34f9303da3c1fd1)
- remove test in version script temporary [`d97929c`](https://github.com/bcomnes/npm-run-all2/commit/d97929cdd73c615e053202c2a5068eb0cf1af08e)

## [v4.1.3](https://github.com/bcomnes/npm-run-all2/compare/v4.1.2...v4.1.3) - 2018-05-07

### Merged

- Fix: unexpected behavior of leading bang [`#124`](https://github.com/bcomnes/npm-run-all2/pull/124)
- Docs: typo fix in README.md: is/are [`#120`](https://github.com/bcomnes/npm-run-all2/pull/120)
- Fix: colorize tasks names sequentially instead of by hash [`#115`](https://github.com/bcomnes/npm-run-all2/pull/115)

### Commits

- Chore: trivial fix [`21c0269`](https://github.com/bcomnes/npm-run-all2/commit/21c02697e59748c0d2ac55b5244bd0ca10ecbfcc)
- Chore: upgrade dependencies [`61bad5c`](https://github.com/bcomnes/npm-run-all2/commit/61bad5c72678f35f8dd1dcd68eb20b688795ffd5)

## [v4.1.2](https://github.com/bcomnes/npm-run-all2/compare/v4.1.1...v4.1.2) - 2017-11-07

### Fixed

- Fix: `--aggregate-output` cannot handle large data (fixes #111)(#112) [`#111`](https://github.com/bcomnes/npm-run-all2/issues/111)

### Commits

- Chore: fix tests [`43a6b16`](https://github.com/bcomnes/npm-run-all2/commit/43a6b1683374e3dd314733ec5559908abfcc6cd1)
- Fix: fix for latest yarn [`dfb9dcb`](https://github.com/bcomnes/npm-run-all2/commit/dfb9dcb40dc0d309480e28dbdaac24878e4c32c6)
- Fix: missing --aggregate-output in npm-run-all [`693261b`](https://github.com/bcomnes/npm-run-all2/commit/693261b6169071b61b443eec5ad6d04525394500)

## [v4.1.1](https://github.com/bcomnes/npm-run-all2/compare/v4.1.0...v4.1.1) - 2017-08-28

### Merged

- Fix: Use ansi-style directly instead of chalk.styles [`#108`](https://github.com/bcomnes/npm-run-all2/pull/108)

### Commits

- Chore: configure appveyor to run on only Node 8. [`ac9358a`](https://github.com/bcomnes/npm-run-all2/commit/ac9358a669611cedb7eb7201e770d60886347ca5)


lib/run-tasks.js
/**
 * @module run-tasks-in-parallel
 * @author Toru Nagashima
 * @copyright 2015 Toru Nagashima. All rights reserved.
 * See LICENSE file in root directory for full license.
 */
"use strict"

//------------------------------------------------------------------------------
// Requirements
//------------------------------------------------------------------------------

const MemoryStream = require("memorystream")
const NpmRunAllError = require("./npm-run-all-error")
const runTask = require("./run-task")

//------------------------------------------------------------------------------
// Helpers
//------------------------------------------------------------------------------

/**
 * Remove the given value from the array.
 * @template T
 * @param {T[]} array - The array to remove.
 * @param {T} x - The item to be removed.
 * @returns {void}
 */
function remove(array, x) {
    const index = array.indexOf(x)
    if (index !== -1) {
        array.splice(index, 1)
    }
}

const signals = {
    SIGABRT: 6,
    SIGALRM: 14,
    SIGBUS: 10,
    SIGCHLD: 20,
    SIGCONT: 19,
    SIGFPE: 8,
    SIGHUP: 1,
    SIGILL: 4,
    SIGINT: 2,
    SIGKILL: 9,
    SIGPIPE: 13,
    SIGQUIT: 3,
    SIGSEGV: 11,
    SIGSTOP: 17,
    SIGTERM: 15,
    SIGTRAP: 5,
    SIGTSTP: 18,
    SIGTTIN: 21,
    SIGTTOU: 22,
    SIGUSR1: 30,
    SIGUSR2: 31,
}

/**
 * Converts a signal name to a number.
 * @param {string} signal - the signal name to convert into a number
 * @returns {number} - the return code for the signal
 */
function convert(signal) {
    return signals[signal] || 0
}

//------------------------------------------------------------------------------
// Public Interface
//------------------------------------------------------------------------------

/**
 * Run npm-scripts of given names in parallel.
 *
 * If a npm-script exited with a non-zero code, this aborts other all npm-scripts.
 *
 * @param {object} tasks - A list of npm-script name to run in parallel.
 * @param {object} options - An option object.
 * @returns {Promise} A promise object which becomes fullfilled when all npm-scripts are completed.
 * @private
 */
module.exports = function runTasks(tasks, options) {
    return new Promise((resolve, reject) => {
        const taskNames = Object.keys(tasks)
        if (taskNames.length === 0) {
            resolve([])
            return
        }

        const results = taskNames.map(taskName => ({ name: taskName, code: undefined }))
        const queue = taskNames.map((taskName, index) => ({ name: taskName, command: tasks[taskName], index }))
        const promises = []
        let error = null
        let aborted = false

        /**
         * Done.
         * @returns {void}
         */
        function done() {
            if (error == null) {
                resolve(results)
            }
            else {
                reject(error)
            }
        }

        /**
         * Aborts all tasks.
         * @returns {void}
         */
        function abort() {
            if (aborted) {
                return
            }
            aborted = true

            if (promises.length === 0) {
                done()
            }
            else {
                for (const p of promises) {
                    p.abort()
                }
                Promise.all(promises).then(done, reject)
            }
        }

        /**
         * Runs a next task.
         * @returns {void}
         */
        function next() {
            if (aborted) {
                return
            }
            if (queue.length === 0) {
                if (promises.length === 0) {
                    done()
                }
                return
            }

            const originalOutputStream = options.stdout
            const optionsClone = Object.assign({}, options)
            const writer = new MemoryStream(null, {
                readable: false,
            })

            if (options.aggregateOutput) {
                optionsClone.stdout = writer
            }

            const task = queue.shift()
            const promise = runTask(task, optionsClone)

            promises.push(promise)
            promise.then(
                (result) => {
                    remove(promises, promise)
                    if (aborted) {
                        return
                    }

                    if (options.aggregateOutput) {
                        originalOutputStream.write(writer.toString())
                    }

                    // Check if the task failed as a result of a signal, and
                    // amend the exit code as a result.
                    if (result.code === null && result.signal !== null) {
                        // An exit caused by a signal must return a status code
                        // of 128 plus the value of the signal code.
                        // Ref: https://nodejs.org/api/process.html#process_exit_codes
                        result.code = 128 + convert(result.signal)
                    }

                    // Save the result.
                    results[task.index].code = result.code

                    // Aborts all tasks if it's an error.
                    if (result.code) {
                        error = new NpmRunAllError(result, results)
                        if (!options.continueOnError) {
                            abort()
                            return
                        }
                    }

                    // Aborts all tasks if options.race is true.
                    if (options.race && !result.code) {
                        abort()
                        return
                    }

                    // Call the next task.
                    next()
                },
                (thisError) => {
                    remove(promises, promise)
                    if (!options.continueOnError || options.race) {
                        error = thisError
                        abort()
                        return
                    }
                    next()
                }
            )
        }

        const max = options.maxParallel
        const end = (typeof max === "number" && max > 0)
            ? Math.min(taskNames.length, max)
            : taskNames.length
        for (let i = 0; i < end; ++i) {
            next()
        }
    })
}
